scylla-proxy/src/proxy.rs

```diff
@@ -0,0 +1,766 @@
+                            }
+                            break
+                        }
+                    }
+                }
+                _terminate = self.terminate_notifier.recv() => break,
+
+                _close = self.connection_close_notifier.recv() => break,
+            }
+        }
+        self.exit("sender_to_cluster");
+    }
+
+    async fn request_processor(
+        mut self,
+        mut requests_rx: mpsc::UnboundedReceiver<RequestFrame>,
+        driver_tx: mpsc::UnboundedSender<ResponseFrame>,
+        cluster_tx: mpsc::UnboundedSender<RequestFrame>,
+        connection_no: usize,
+        request_rules: Arc<RwLock<Vec<RequestRule>>>,
+        connection_close_signaler: ConnectionCloseSignaler,
+    ) {
+        'mainloop: loop {
+            tokio::select! {
+                request = requests_rx.recv() => {
+                    match request {
+                        Some(request) => {
+                            let ctx = EvaluationContext{
+                                connection_seq_no: connection_no,
+                                opcode: FrameOpcode::Request(request.opcode),
+                                frame_body: request.body.clone(),
+                            };
+                            let guard = request_rules.read().await;
+                '_ruleloop: for RequestRule(condition, reaction) in guard.iter() {
+                                if condition.eval(&ctx) {
+                                    let cluster_tx_clone = cluster_tx.clone();
+                                    let request_clone = request.clone();
+                                    let pass_action = async move {
+                                        if let Some(ref pass_action) = reaction.to_addressee {
+                                            if let Some(time) = pass_action.delay {
+                                                tokio::time::sleep(time).await;
+                                            }
+                                            let passed_frame = match pass_action.msg_processor {
+                                                Some(ref processor) => processor(request_clone),
+                                                None => request_clone,
+                                            };
+                                            let _ = cluster_tx_clone.send(passed_frame);
+                                        };
+                                    };
+
+                                    let driver_tx_clone = driver_tx.clone();
+                                    let request_clone = request.clone();
+                                    let forge_action = async move {
+                                        if let Some(ref forge_action) = reaction.to_sender {
+                                            if let Some(time) = forge_action.delay {
+                                                tokio::time::sleep(time).await;
```

Comment:
Do I understand correctly that this will block processing other frames until the sleep finishes?

Reply:
Ah, it's fixed in a later commit. Still, it's confusing when reading commit-by-commit.